onmt/translate/translator.py

```diff
@@ -715,12 +626,14 @@ def _translate_batch(
                 if src_map is not None:
                     src_map = src_map.index_select(1, select_indices)
 
-            self.model.decoder.map_state(
-                lambda state, dim: state.index_select(dim, select_indices))
+            if strategy.update_state:
```

Comment:
Can this be detected automatically? Maybe something like `not torch.equal(select_indices, torch.range(current_batch_size))`?

Reply:
I don't think this can be detected automatically.
Previously, we want to execute `.map_state` after each `.advance` for beam search or each `.update_finished` for random sampling.
I've tried to use a condition like `not torch.equal(select_indices, previous_select_indices)`,
this does not work as we may assign `select_indices` with exactly the same value.

Reply:
I think comparing to `torch.range` should work as it means "batch size is the same and no indices changed". It's just that I would prefer an alternative to communicating with the `update_state` attribute.

Reply:
The `torch.range` thing indeed works in the case of beam search, but not in the case of random sampling. I could not find a concise condition that could work for both. As an alternative to `update_state`, maybe we still need to separately condition on these two case.
Here is my proposition:
```
parallel_paths = decode_strategy.is_finished.shape[1]  # the beam_size
Loop step{
    ....
    if parallel_paths >  1 or any_finished:  # beam search or (random sampling & any_batch_finished)
         self.model.decoder.map_state(...)
}
```

Reply:
The issue is that `RandomSampling` does not update `select_indices` in `advance` like `BeamSearch` does. This should be better enforced by the interface otherwise client codes have to guess how and when to use `select_indices`. Your proposition is OK for now.